pkg/codegen/internal/test/testdata/cyclic-types/go/example/pulumiTypes.go

```diff
@@ -0,0 +1,581 @@
+// You can construct a concrete instance of `IndirectCycleSPtrInput` via:
+//
+//          IndirectCycleSArgs{...}
+//
+//  or:
+//
+//          nil
+type IndirectCycleSPtrInput interface {
+	pulumi.Input
+
+	ToIndirectCycleSPtrOutput() IndirectCycleSPtrOutput
+	ToIndirectCycleSPtrOutputWithContext(context.Context) IndirectCycleSPtrOutput
+}
+
+type indirectCycleSPtrType IndirectCycleSArgs
+
+func IndirectCycleSPtr(v *IndirectCycleSArgs) IndirectCycleSPtrInput {
+	return (*indirectCycleSPtrType)(v)
+}
+
+func (*indirectCycleSPtrType) ElementType() reflect.Type {
+	return reflect.TypeOf((**IndirectCycleS)(nil)).Elem()
+}
+
+func (i *indirectCycleSPtrType) ToIndirectCycleSPtrOutput() IndirectCycleSPtrOutput {
+	return i.ToIndirectCycleSPtrOutputWithContext(context.Background())
+}
+
+func (i *indirectCycleSPtrType) ToIndirectCycleSPtrOutputWithContext(ctx context.Context) IndirectCycleSPtrOutput {
+	return pulumi.ToOutputWithContext(ctx, i).(IndirectCycleSPtrOutput)
+}
+
+type IndirectCycleSOutput struct{ *pulumi.OutputState }
+
+func (IndirectCycleSOutput) ElementType() reflect.Type {
+	return reflect.TypeOf((*IndirectCycleS)(nil)).Elem()
+}
+
+func (o IndirectCycleSOutput) ToIndirectCycleSOutput() IndirectCycleSOutput {
+	return o
+}
+
+func (o IndirectCycleSOutput) ToIndirectCycleSOutputWithContext(ctx context.Context) IndirectCycleSOutput {
+	return o
+}
+
+func (o IndirectCycleSOutput) ToIndirectCycleSPtrOutput() IndirectCycleSPtrOutput {
+	return o.ToIndirectCycleSPtrOutputWithContext(context.Background())
+}
+
+func (o IndirectCycleSOutput) ToIndirectCycleSPtrOutputWithContext(ctx context.Context) IndirectCycleSPtrOutput {
+	return o.ApplyTWithContext(ctx, func(_ context.Context, v IndirectCycleS) *IndirectCycleS {
+		return &v
+	}).(IndirectCycleSPtrOutput)
+}
+
+func (o IndirectCycleSOutput) Foo2() IndirectCycleTPtrOutput {
+	return o.ApplyT(func(v IndirectCycleS) *IndirectCycleT { return v.Foo2 }).(IndirectCycleTPtrOutput)
+}
+
+type IndirectCycleSPtrOutput struct{ *pulumi.OutputState }
+
+func (IndirectCycleSPtrOutput) ElementType() reflect.Type {
+	return reflect.TypeOf((**IndirectCycleS)(nil)).Elem()
+}
+
+func (o IndirectCycleSPtrOutput) ToIndirectCycleSPtrOutput() IndirectCycleSPtrOutput {
+	return o
+}
+
+func (o IndirectCycleSPtrOutput) ToIndirectCycleSPtrOutputWithContext(ctx context.Context) IndirectCycleSPtrOutput {
+	return o
+}
+
+func (o IndirectCycleSPtrOutput) Elem() IndirectCycleSOutput {
+	return o.ApplyT(func(v *IndirectCycleS) IndirectCycleS {
+		if v != nil {
+			return *v
+		}
+		var ret IndirectCycleS
+		return ret
+	}).(IndirectCycleSOutput)
+}
+
+func (o IndirectCycleSPtrOutput) Foo2() IndirectCycleTPtrOutput {
+	return o.ApplyT(func(v *IndirectCycleS) *IndirectCycleT {
+		if v == nil {
+			return nil
+		}
+		return v.Foo2
+	}).(IndirectCycleTPtrOutput)
+}
+
+type IndirectCycleT struct {
+	Foo3 *IndirectCycleS `pulumi:"foo3"`
```

Comment:
both IndirectCycleS and IndirectCycleT properties get '*' treatment, so this is symmetric/deterministic, good!